bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -440,7 +440,7 @@ STDigest BCStateTran::checkpointReservedPages(uint64_t checkpointNumber, DataSto
 // Remove old checkpoints from the data store
 void BCStateTran::deleteOldCheckpoints(uint64_t checkpointNumber, DataStoreTransaction *txn) {
   uint64_t minRelevantCheckpoint = 0;
-  if (checkpointNumber > maxNumOfStoredCheckpoints_) {
+  if (checkpointNumber >= maxNumOfStoredCheckpoints_) {
```

Comment:
I fail to understand how this makes any difference, in light of the fact that right below, we ignore `minRelevantCheckpoint < 2`. With 1minRelevantCheckpoint = 3` and `checkpointNumber = 4`, this should still only store checkpoints 2-4, which is 3 checkpoints. 

I think something more complicated is going on here. It's almost like the transaction isn't working as expected.

Reply:
Hi Andrew, thanks for looking into this. I believe the problem is actually quite simple and it is indeed a corner case. 
Let's put this into concrete terms. When the assert fails, we have the following state:
`2020-03-27T10:53:32.975 [139864041720448] INFO  state-transfer %% ********* firstStoredCheckpoint = 0, lastStoredCheckpoint = 3, psd_->hasCheckpointDesc(firstStoredCheckpoint) = 0, maxNumOfStoredCheckpoints_ = 3`

In this case indeed, `lastStoredCheckpoint - firstStoredCheckpoint + 1 = 4 > 3 = maxNumOfStoredCheckpoints_` which is a failure of the invariant.

This "failed" state is possible because of the "if" statement I'm fixing in this PR. If we keep the strict inequality `>` at line 443, then `minRelevantCheckpoint=0` which means that `firstStoredCheckpoint=0` remains and the 0-th checkpoint is not deleted.

To summarize - it's not an issue with the transaction, but rather with the logic that deletes old checkpoints.

Reply:
Aha. I didn't consider that firstStoredCheckpoint could equal 0. Thank you for pointing that out Teo. So what we really are doing here is ensuring that we are setting `firstStoredCheckpoint` to 1 immediately. 

I find this logic confusing with a few to many conditionals. Nice job tracking this down and fixing it.  There really is no such thing as a 0th stored checkpoint which is why this is so confusing. It's a failure of the unsigned type having a bottom value that is a valid number, but not really a valid checkpoint. Using a `std::optional` would have made this explicit.